### PR TITLE
Resolve types properly in const eval

### PR DESCRIPTION
**Pull request by oli-obk**
r? @eddyb 

cc @arielb1

### CODE REVIEW

**Comment by eddyb** [comment]
Can you add a test, with an associated `const` forwarding to another, generically?
That sort of thing wouldn't work without this change.

**Comment by oli-obk** [comment]
> Can you add a test, with an associated const forwarding to another, generically?
That sort of thing wouldn't work without this change.

I tried. But it was a silent error before, and it works now, so there's no difference. Using the assoc constant won't work generically out of other reasons: https://play.rust-lang.org/?gist=54e162b150e745236878f2f0eca06de3&version=stable

**Comment by eddyb** [comment]
@oli-obk There's no forwarding going on there between two impls' associated `const`s - and you can't have array lengths that depend on type parameters yet.

Closer to what I meant: https://play.rust-lang.org/?gist=e0ec487c9566ea8f8187eb14d82e62ac&version=nightly

File: src/librustc_const_eval/eval.rs [hunk 1]
```diff
@@ -289,6 +267,7 @@ fn eval_const_expr_partial<'a, 'tcx>(cx: &ConstContext<'a, 'tcx>,
           match cx.tables.qpath_def(qpath, e.hir_id) {
               Def::Const(def_id) |
               Def::AssociatedConst(def_id) => {
+                    let substs = tcx.normalize_associated_type_in_env(&substs, cx.param_env);
```
Comment (arielb1) [inline src/librustc_const_eval/eval.rs]: Isn't this normalization also needed in the `Def::Method(id) | Def::Fn(id)` arm?

**Comment by arielb1** [comment]
Another test for broken subst logic (this shouldn't ICE):
```Rust
#![allow(unused)]

trait Foo {
    const BAR: Self;
}

impl Foo for i32 {
    const BAR: Self = 4;
}
impl Foo for i64 {
    const BAR: Self = 3;
}

struct Bar<T: ?Sized>(usize, std::marker::PhantomData<T>);

impl<T: ?Sized> Foo for Bar<T> {
    const BAR: Self = Bar(4, std::marker::PhantomData);
}

impl<T: ?Sized> Bar<T> {
    fn foo() {
        let x = Self::BAR.0;
        let x: &'static usize = &Self::BAR.0;
        let x: [i32; Self::BAR.0] = [1, 2];
    }
}

fn main() {
    let x: [i32; i32::BAR as usize] = [1, 2, 3, 4];
    let x: [i32; i64::BAR as usize] = [1, 2, 3];
}
```

r+ with proper tests

**Comment by oli-obk** [comment]
@arielb1 That subst doesn't come from const eval, it comes from https://github.com/rust-lang/rust/blob/master/src/librustc/traits/project.rs#L357

File: src/test/run-pass/ctfe/assoc-const.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(const_size_of)]
+#![allow(unused)]
```
Comment (eddyb) [inline src/test/run-pass/ctfe/assoc-const.rs]: You should remove this.

File: src/test/run-pass/ctfe/assoc-const.rs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(const_size_of)]
```
Comment (eddyb) [inline src/test/run-pass/ctfe/assoc-const.rs]: Wait, this too, I think? Since the example doesn't use `size_of` anymore.

Comment (oli-obk) [reply src/test/run-pass/ctfe/assoc-const.rs]: done

**Comment by oli-obk** [comment]
travis likes it + comments addressed

File: src/librustc_const_eval/eval.rs (hunk 1 above)
Comment (oli-obk) [reply src/librustc_const_eval/eval.rs]: technically we only need it for `Def::AssociatedConst`, but it's a nop for `Def::Const`

**Pull request merged**